src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4

```diff
@@ -768,11 +768,29 @@ template
     ;
 
 stringTemplate
-    : STRING_TEMPLATE_BEGIN expr? stringTemplateMiddle* STRING_TEMPLATE_END
+    : stringTemplateBegin expr? stringTemplateMiddle* stringTemplateEnd
     ;
 
 stringTemplateMiddle
-    : STRING_TEMPLATE_MID expr?
+    : stringTemplateMid expr?
+    ;
+
+stringTemplateBegin
+    : STRING_TEMPLATE_BEGIN
+      STRING_TEMPLATE_CONTENT?
+      EMBEDDED_EXPRESSION_BEGIN
+    ;
+
+stringTemplateMid
+    : EMBEDDED_EXPRESSION_END
+      STRING_TEMPLATE_CONTENT?
+      EMBEDDED_EXPRESSION_BEGIN
+    ;
+
+stringTemplateEnd
+    : EMBEDDED_EXPRESSION_END
+      STRING_TEMPLATE_CONTENT?
+      STRING_TEMPLATE_END
     ;
```

Comment:
These parser rules are now more reflective of what's in the JLS: https://docs.oracle.com/javase/specs/jls/se21/preview/specs/string-templates-jls.html#:~:text=template%20(3.13).-,3.13%20Fragments,-A%20template%20(